pkg/queue/queue.go

```diff
@@ -0,0 +1,210 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+
+	"istio.io/pkg/log"
+)
+
+// Queue of work tickets processed using a rate-limiting loop
+type Queue interface {
+	// Push a task.
+	Push(item interface{}, task Task)
+	// Run the loop until a signal on the channel
+	Run(<-chan struct{})
+}
+
+type (
+	empty struct{}
+	t     interface{}
+	set   map[t]empty
+)
+
+func (s set) has(item t) bool {
+	_, exists := s[item]
+	return exists
+}
+
+func (s set) insert(item t) {
+	s[item] = empty{}
+}
+
+func (s set) delete(item t) {
+	delete(s, item)
+}
+
+type innerTask struct {
+	item interface{}
+	*backoff.ExponentialBackOff
+	task Task
+}
+
+type BackOffOption struct {
+	// Initial backoff interval.
+	InitialInterval time.Duration
+	// Max Backoff interval.
+	MaxInterval time.Duration
+	// After MaxElapsedTime the ExponentialBackOff stops.
+	// It never stops if MaxElapsedTime == 0.
+	MaxElapsedTime time.Duration
+}
+
+type queue struct {
+	backoffOpts BackOffOption
+	tasks       []*innerTask
+	// Things that are currently being processed are in the processing set.
+	// These things may be simultaneously in the dirty set. When we finish
+	// processing something and remove it from this set, we'll check if
+	// it's in the dirty set, and if so, add it to the queue.
+	processing set
+	// dirty defines all of the items that need to be processed.
+	// it has the latest task.
+	dirty   map[interface{}]*innerTask
+	cond    *sync.Cond
+	closing bool
+}
+
+func NewBackOffQueue(backoff BackOffOption) Queue {
+	return &queue{
+		backoffOpts: backoff,
+		tasks:       make([]*innerTask, 0),
+		processing:  make(map[t]empty),
+		dirty:       make(map[interface{}]*innerTask),
```

Comment:
it's wired to see that using `t` as `interface{}` in `processing`, but using `interface{}` in `dirty`. Or just make `processing: make(set)`?

Reply:
Need the latest task recorded in dirty